File: algorithms/cracking_the_coding_interview/queues_a_tale_of_two_stacks.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

const (
	TYPE_ENQUEUE = 1
	TYPE_DEQUEUE = 2
	TYPE_PEEK    = 3
)

type Queue struct {
	stackInput  []int
	stackOutput []int
}

func (queue *Queue) Enqueue(value int) {
	queue.stackInput = append(queue.stackInput, value)
}

func (queue *Queue) Dequeue() {
	if len(queue.stackOutput) == 0 {
		queue.moveInputToOutput()
	}

	length := len(queue.stackOutput)
	if length == 0 {
		return
	}

	queue.stackOutput = queue.stackOutput[:length-1]
}

func (queue *Queue) Peek() (int, error) {
	if len(queue.stackOutput) == 0 {
		queue.moveInputToOutput()
	}

	length := len(queue.stackOutput)
	if length == 0 {
		return 0, fmt.Errorf("nothing is returned")
	}

	return queue.stackOutput[length-1], nil
}

func (queue *Queue) moveInputToOutput() {
	length := len(queue.stackInput)
	for i := length - 1; i >= 0; i-- {
		queue.stackOutput = append(queue.stackOutput, queue.stackInput[i])
	}
	queue.stackInput = []int{}
}

func process(queue *Queue, typeId int, typeValue int) (int, error) {
	if typeId == TYPE_ENQUEUE {
		queue.Enqueue(typeValue)
		return 0, fmt.Errorf("nothing is returned")
	}

	if typeId == TYPE_DEQUEUE {
		queue.Dequeue()
		return 0, fmt.Errorf("nothing is returned")
	}

	if typeId == TYPE_PEEK {
		return queue.Peek()
	}

	return 0, fmt.Errorf("nothing is returned")
}

func main() {
	reader := bufio.NewReader(os.Stdin)
	reader.ReadString('\n')
	queue := Queue{}

	for {
		line, err := reader.ReadString('\n')
		line = strings.TrimSpace(line)
		lineData := strings.Fields(line)

		typeId, _ := strconv.Atoi(lineData[0])
		typeValue := 0
		if len(lineData) > 1 {
			typeValue, _ = strconv.Atoi(lineData[1])
		}

		if value, err := process(&queue, typeId, typeValue); err == nil {
			fmt.Println(value)
		}

		if err != nil {
			break
		}
	}
}

File: algorithms/cracking_the_coding_interview/queues_a_tale_of_two_stacks.php
<?php
declare(strict_types=1);

const TYPE_ENQUEUE = 1;
const TYPE_DEQUEUE = 2;
const TYPE_PEEK = 3;

class Queue
{
    private $stackInput = [];
    private $stackOutput = [];

    /**
     * @param int $value
     * @return void
     */
    public function enqueue(int $value)
    {
        $this->stackInput[] = $value;
    }

    /**
     * @return void
     */
    public function dequeue()
    {
        if (empty($this->stackOutput)) {
            $this->moveInputToOutput();
        }

        $length = count($this->stackOutput);
        if ($length === 0) {
            return;
        }

        array_pop($this->stackOutput);
    }

    /**
     * @return int|null
     */
    public function peek()
    {
        if (empty($this->stackOutput)) {
           $this->moveInputToOutput();
        }

        $length = count($this->stackOutput);
        if ($length === 0) {
            return null;
        }

        return $this->stackOutput[$length - 1];
    }

    /**
     * @return void
     */
    private function moveInputToOutput()
    {
        $length = count($this->stackInput);
        for ($i = $length - 1; $i >= 0; $i--) {
            $this->stackOutput[] = $this->stackInput[$i];
        }
        $this->stackInput = [];
    }
}

/**
 * @param Queue $queue
 * @param int $type
 * @param int|null $typeValue
 * @return mixed
 */
function process(Queue $queue, int $type, int $typeValue = null)
{
    if ($type === TYPE_ENQUEUE && !is_null($typeValue)) {
        $queue->enqueue($typeValue);
        return null;
    }

    if ($type === TYPE_DEQUEUE) {
        $queue->dequeue();
        return null;
    }

    if ($type === TYPE_PEEK) {
        return $queue->peek();
    }

    return null;
}

$handle = fopen('php://stdin', 'r');
fscanf($handle, '%d', $t);

$result = [];
$queue = new Queue();
for ($a0 = 0; $a0 < $t; $a0++) {
    fscanf($handle, '%d %d', $type, $typeValue);

    $value = process($queue, $type, $typeValue);
    if (!is_null($value)) {
        $result[] = $value;
    }
    // not to reuse value from a previous loop
    $typeValue = null;
}
echo implode("\n", $result);
